pkg/console/controllers/plugins/controller.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/openshift/console-operator/pkg/console/status"
+)
+
+const (
+	MigratedPluginsAnnotation = "console.openshift.io/migrated-plugins"
+)
+
+type PluginsMigrationController struct {
+	// configs
+	operatorClient       v1helpers.OperatorClient
+	operatorConfigClient operatorclientv1.ConsoleInterface
+	// lister
+	consolePluginLister   listerv1alpha1.ConsolePluginLister
+	consolePluginInformer consoleinformersv1alpha1.ConsolePluginInformer
+}
+
+func NewPluginsMigrationController(
+	// operator
+	operatorClient v1helpers.OperatorClient,
+	operatorConfigClient operatorclientv1.ConsoleInterface,
+	// operatorConfigInformer operatorinformerv1.ConsoleInformer,
+	// plugins
+	consolePluginInformer consoleinformersv1alpha1.ConsolePluginInformer,
+	// events
+	recorder events.Recorder,
+) factory.Controller {
+
+	c := &PluginsMigrationController{
+		// configs
+		operatorClient:       operatorClient,
+		operatorConfigClient: operatorConfigClient,
+		// plugins
+		consolePluginLister: consolePluginInformer.Lister(),
+	}
+
+	return factory.New().
+		WithInformers(
+			consolePluginInformer.Informer(),
+		).WithSync(c.Sync).
+		ToController("PluginsMigrationController", recorder.WithComponentSuffix("plugins-migration-controller"))
+}
+
+func (c *PluginsMigrationController) Sync(ctx context.Context, controllerContext factory.SyncContext) error {
+
+	operatorConfig, err := c.operatorConfigClient.Get(ctx, api.ConfigResourceName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	switch operatorConfig.Spec.ManagementState {
+	case operatorv1.Managed:
+		klog.V(4).Infof("console-operator is in a managed state: migrating available plugins")
+	case operatorv1.Unmanaged:
+		klog.V(4).Infof("console-operator is in an unmanaged state: skipping available plugins migration")
+		return nil
+	case operatorv1.Removed:
+		klog.V(4).Infof("console-operator is in a removed state: removing migrated plugins")
+		return c.removeMigratedPlugins(ctx, operatorConfig)
+	default:
+		return fmt.Errorf("unknown state: %v", operatorConfig.Spec.ManagementState)
+	}
+
+	statusHandler := status.NewStatusHandler(c.operatorClient)
+
+	availablePluginsName, err := c.GetAvailablePluginsName()
+	statusHandler.AddConditions(status.HandleProgressingOrDegraded("AvailablePluginMigration", "FailedAvailablePluginsGet", err))
+	if err != nil {
+		return statusHandler.FlushAndReturn(err)
+	}
+	migratedPluginsArray, err := GetMigratedPlugins(operatorConfig)
+	statusHandler.AddConditions(status.HandleProgressingOrDegraded("AvailablePluginMigration", "FailedMigratedPluginsGet", err))
+	if err != nil {
+		return statusHandler.FlushAndReturn(err)
+	}
+
+	pluginsToEnable := GetPluginsToEnable(operatorConfig, availablePluginsName, migratedPluginsArray)
+
+	if len(pluginsToEnable) == 0 {
+		return statusHandler.FlushAndReturn(nil)
+	}
+
+	err = c.UpdateOperatorConfig(ctx, operatorConfig, pluginsToEnable, migratedPluginsArray)
```

Comment:
I think we want to add the well-known plugins to the migrated plugins array even if we didn't enable them. If kubevirt isn't already installed before the console upgrade, we don't want to automatically enable it when it's installed later. The use case we're trying to avoid is disabling the kubevirt UI when it was previously enabled in 4.8.

So

1. If `kubevirt` ConsolePlugin exists, automatically enable it.
2. Add `kubevrit` to the list of migrated plugins (even if the ConsolePlugin wasn't found).

I'm assuming here the kubevirt operator with the plugin must be updated before upgrading the cluster. We'll need to check with the kubevirt team. cc @glekner @vojtechszocs